### PR TITLE
Yellow slime cores now have a more reasonable recharge rate

### PR DESCRIPTION
**Pull request by Rukofamicom**
## About The Pull Request

It reduces the ludicrous rate at which charged and hypercharged slime cores regenerate power, and also lowers their capacity to make them a unique trade-off instead of a direct upgrade. EMPs now block the regeneration for 30 seconds, so that if one is drained to zero, it isn't back to having charge the next second. 

### Exact values changed:
Normal Yellow:
Regeneration: 375/s -> 25/s
Max Charge: 10k -> 5k
It is suitable for APCs and cyborgs which do not experience high drain. Not recommended for mechs and negligible effect in SMES

Hypercharged Yellow:
Regeneration: 562/s -> 75/s
Max Charge: 20k -> 10k
Suitable for all purposes and has a charge rate that feels decent, even if not infinite like it used to be. Still won't notice them much in SMES, but they will help somewhat. 

## Why It's Good For The Game

Yellow slime cores do not provide a power cell that recharges itself over time, they provide a a power cell with infinite power that stays full almost all the time unless you're using the heaviest power drains available. Said power drains have such high costs to limit their number of uses, and slime cores totally negate this.

A normal yellow core currently recharges at a rate of 375 per second, and hypercharged at a rate of 562.5. Compare this to the costs of tools on Cyborgs and mechs

I didn't do the exact math because when it came to APCs and SMES, even normal slime cores always produced more power than I could possibly get to drain from either. 

## Testing Photographs and Procedure

Tested EMP mechanics with (now deleted) debug messages to ensure regeneration was halting when expected and resuming when expected. Tested in APCs, SMES and Cyborgs, though I did not extensively test beyond EMPs and ensuring that recharging was occurring since this is just a value adjustment. 

image
Showing debug messages from when I realized delta_time is in seconds instead of deciseconds. 

## Changelog
:cl:
balance: Yellow slime cores are now cells that recharge over time instead of being effectively infinite.
balance: EMPs now halt regeneration for 30 seconds on slime cores. 
/:cl:

### CODE REVIEW

**Comment by ToasterBan** [comment]
might want to do the same with botany plants cells espiecally since they can become T4 equivelants

**Comment by Rukofamicom** [comment]
Botany power cells don't regenerate their charge though. Only two power cells have the self-charging capability variable set to true.

**Comment by EvilDragonfiend** [comment]
> might want to do the same with botany plants cells espiecally since they can become T4 equivelants

Wait a bit if you are concerned. I am working on botany rework.

**Comment by williamcll** [comment]
> might want to do the same with botany plants cells espiecally since they can become T4 equivelants

They don't recharge. And having T4 cell production being reliant on one department is awful.

File: code/modules/power/cell.dm [hunk 1]
```diff
@@ -11,15 +11,24 @@
...
 	grind_results = list(/datum/reagent/lithium = 15, /datum/reagent/iron = 5, /datum/reagent/silicon = 5)
-	var/rigged = FALSE	// true if rigged to explode
-	var/chargerate = 100 //how much power is given every tick in a recharger
-	var/self_recharge = 0 //does it self recharge, over time, or not?
+	/// true if rigged to explode
+	var/rigged = FALSE	
+	///how much power is given every tick in a recharger
+	var/chargerate = 100 
+	///does it self recharge, over time, or not?
+	var/self_recharge = 0
```
Comment (ike709) [inline code/modules/power/cell.dm]:
```suggestion
	var/self_recharge = FALSE
```

File: code/modules/power/cell.dm [hunk 2]
```diff
@@ -348,6 +369,8 @@
 	materials = list()
 	rating = 5 //self-recharge makes these desirable
 	self_recharge = 1 // Infused slime cores self-recharge, over time
```
Comment (ike709) [inline code/modules/power/cell.dm]:
```suggestion
	self_recharge = TRUE // Infused slime cores self-recharge, over time
```

**Review by ike709 — CHANGES_REQUESTED** [review]
Only checked the code not the balance. At a glance it seems fine though?

Also add the EMP stacking to the CL. It also might be worth capping how high the EMPs can stack to prevent seemingly endless EMP stacking.

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]
Use a world.time timer for the EMP rather than a counter.

File: code/modules/power/cell.dm [hunk 3]
```diff
@@ -11,15 +11,24 @@
...
-	var/self_recharge = 0 //does it self recharge, over time, or not?
+	/// true if rigged to explode
+	var/rigged = FALSE	
+	///how much power is given every tick in a recharger
+	var/chargerate = 100 
+	///does it self recharge, over time, or not?
+	var/self_recharge = FALSE
+	///stores the chargerate to restore when hit with EMP, for slime cores
+	var/emp_chargerate_save
+	///Timer for EMPs so that recharging cells know when to start charging again.
```
Comment (ike709) [inline code/modules/power/cell.dm]:
```suggestion
	///Time in SECONDS for EMPs so that recharging cells know when to start charging again. 
```

File: code/modules/power/cell.dm [hunk 4]
```diff
@@ -136,6 +150,12 @@
 	charge -= 1000 / severity
 	if (charge < 0)
 		charge = 0
+	if(self_recharge)
+		if(chargerate)
+			emp_chargerate_save = chargerate
+			chargerate = 0
+		emp_timer = 30		//30 seconds
```
Comment (ike709) [inline code/modules/power/cell.dm]:
```suggestion
		emp_timer = 30		// emp_timer is in seconds
```

File: code/modules/power/cell.dm [hunk 5]
```diff
@@ -53,6 +62,11 @@
 		give(chargerate * 0.125 * delta_time)
 	else
 		return PROCESS_KILL
+	if(emp_timer)
+		emp_timer -= delta_time
```
Comment (ike709) [inline code/modules/power/cell.dm]:
```suggestion
		emp_timer -= delta_time // Both vars are in seconds
```

**Comment by ike709** [comment]
Probably fine, leaving this in bacon's hands.

**Comment by Rukofamicom** [comment]
I re-tested everything again after completely changing how the EMP timer is handled.

As far as the balance side of things:

Cyborgs using either core, even the weaker one, will still have *mostly* limitless power unless they're aggressively spamming something or another. For most typical gameplay*, there will not be a noticeable difference in functionality. As for the asterisk, that's to cover the engiborg RCD. Borg RCD has a very high cost and can be adjusted in a separate PR if necessary - I think the cost is high on purpose because the borg RCD is completely free to use outside of cell charge; it does not require or drain the station's physical resources in any way. 

As a general point for comparison: 
* Cyborg light on full power drains 12 charge per cycle. Normal core provides more than double this. 
* Self-repair module (before my other PR) is considered a fairly punishing drain to leave on all the time - at its worst state it drains 7.5 charge per cycle.
* Living cyborgs that are not stunned passive drain 1 energy per cycle. 

With the basic core (25 energy per cycle) you can have both of these running and still have a net positive gain on energy. The reduction in capacity only really serves to exacerbate the EMP weakness a little bit.

Mechs I didn't explicitly factor or balance around, but all of the non-infinite cores remain just as available as before, and if mechs were only worth using with an infinite energy source, I'd say they have entirely separate issues to resolve.

**Comment by Tsar-Salat** [comment]
> Mechs I didn't explicitly factor or balance around, but all of the non-infinite cores remain just as available as before, and if mechs were only worth using with an infinite energy source, I'd say they have entirely separate issues to resolve.

With this, I agree.
I've found that the Gygax's unique ability drains the battery  to half with a simple walk from Sec to Science with a full recharge from the yellow taking around 20 seconds for the hyper and probably twice that for regular.

Now this sounds pretty bad for balancing reasons, but changing it to be even *less* than it currently is basically means there's no point in using the Gyagax ability to try to keep up and pacify people, but just use long range weapons to Tesla rifle and then spam laser.

**Comment by Rukofamicom** [comment]
As more or less implied in this PR, I don't have the experience with mechs to confidently try rebalancing costs, but I can say it will be quite easy to make a that simply adjusts the costs if you want to take a crack at fixing that issue with your experience behind it.

**Comment by PerishedFraud** [comment]
Not opposed to the change but it does feel a bit sour nerfing one of the few things xenobio can make that aren't powergamey and benefit the whole station without outclassing other jobs.

Neutral vote tbh. If i wanted to nerf xenobio i'd focus on some of their other bullshit, so it does feel wrong seeing the nice guy option getting nerfed. On the other hand this is a fallacy. The existence of worse shit doesn't make the cores any less stupid.

Shrug.

File: code/modules/power/cell.dm [hunk 6]
```diff
@@ -11,15 +11,24 @@
...
+	/// true if rigged to explode
+	var/rigged = FALSE	
+	///how much power is given every tick in a recharger
+	var/chargerate = 100 
+	///does it self recharge, over time, or not?
+	var/self_recharge = FALSE
+	///stores the chargerate to restore when hit with EMP, for slime cores
+	var/emp_chargerate_save
+	///Time in SECONDS for EMPs so that recharging cells know when to start charging again. 
+	var/emp_timer = 0
```
Comment (PowerfulBacon) [inline code/modules/power/cell.dm:26]: Rather than using a timer that needs updating every tick, use a world.time counterdown.
There are some defines for cooldown_initialize, but its pretty simple to write one quickly anyway.

Simply have the timer represent a world time of how long the battery should be disabled until. If the current world time is less than that value, do nothing.

To set the battery on a cooldown you can do emp_timer = world.time + 30 SECONDS or whatever you want the cooldown to be.
To check if the battery is still on cooldown you can do
```
if(emp_timer < world.time)
  return
```

**Comment by Rukofamicom** [comment]
Testing

**Comment by Rukofamicom** [comment]
Okay something is fucky, but it's not working in a **bad** way. 
1) EMP hits power cell
2) Powercell drains
3) Powercell gets appropriate timer if self-charging
4) Breakpoint triggers to cancel self charge for a while
5) Breakpoint triggers indicating that recharging has started
6) Process() starts getting called less frequently for some reason ????
7) After about 30 more seconds Process() starts getting called every interval it's expected to again ????

To summarize:
image

## It technically works as intended, with an unintended but kinda cool side effect of charge being slower for a short while when it starts charging again. I have no idea why. 

This extra effect did *not* exist before now.

**Review by PowerfulBacon — APPROVED** [review]

**Review by ike709 — APPROVED** [review]
